parley/src/layout/mod.rs

```diff
@@ -261,3 +267,149 @@ pub struct Decoration<B: Brush> {
     /// containing run.
     pub size: Option<f32>,
 }
+
+#[cfg(feature = "accesskit")]
+#[derive(Clone, Default)]
+pub struct LayoutAccessibility {
+    // The following two fields maintain a two-way mapping between runs
+    // and AccessKit node IDs, where each run is identified by its line index
+    // and run index within that line, or a run path for short. These maps
+    // are maintained by `LayoutAccess::build_nodes`, which ensures that removed
+    // runs are removed from the maps on the next accessibility pass.
+    pub(crate) access_ids_by_run_path: HashMap<(usize, usize), NodeId>,
+    pub(crate) run_paths_by_access_id: HashMap<NodeId, (usize, usize)>,
+}
+
+#[cfg(feature = "accesskit")]
+impl LayoutAccessibility {
+    #[allow(clippy::too_many_arguments)]
+    pub fn build_nodes<B: Brush>(
+        &mut self,
+        text: &str,
+        layout: &Layout<B>,
+        update: &mut TreeUpdate,
+        parent_node: &mut Node,
+        mut next_node_id: impl FnMut() -> NodeId,
+        x_offset: f64,
+        y_offset: f64,
+    ) {
+        // Build a set of node IDs for the runs encountered in this pass.
+        let mut ids = HashSet::<NodeId>::new();
+
+        for (line_index, line) in layout.lines().enumerate() {
+            let metrics = line.metrics();
+            // Defer adding each run node until we reach either the next run
+            // or the end of the line. That way, we can set relations between
+            // runs in a line and do anything special that might be required
+            // for the last run in a line.
+            let mut last_node: Option<(NodeId, Node)> = None;
+
+            // Iterate over the runs from left to right, computing their offsets,
+            // then sort them into text order.
+            let runs = {
+                let mut run_offset = metrics.offset;
+                let mut runs = Vec::with_capacity(line.len());
```

Comment:
Should we reuse this vector? Maybe not between calls, but at least between loop iterations.